File: test_lwjgl3/src/Engine/Window.java
package Engine;

import com.generic.graphics.render.FPSCounter;
import org.joml.Matrix4f;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.glfw.GLFWVidMode;
import org.lwjgl.opengl.GL;

import static org.lwjgl.glfw.GLFW.*;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL11C.glViewport;
import static org.lwjgl.system.MemoryUtil.*;

public class Window {
    private long glfwWindow;
    private int width;
    private int height;
    private String title;
    private boolean resized;
    private boolean vSync;

    private FPSCounter fps;

    public static final float FOV = (float) Math.toRadians(60.0f);

    public static final float Z_NEAR = 0.01f;

    public static final float Z_FAR = 1000.0f;

    private Matrix4f projectionMatrix;

    public Window(int width, int height, String title, boolean vSync)
    {
        this.width = width;
        this.height = height;
        this.title = title;
        this.vSync = vSync;
        this.resized = false;
    }

    public void init()
    {
        //retour d'erreur de glfw dans la console.
        GLFWErrorCallback.createPrint(System.err).set();

        if (!glfwInit())
        {
            throw new IllegalStateException("impossible d'initialiser GLFW");
        }

        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_VISIBLE, GL_FALSE);                             //fenêtre masquée
        glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);                            //fenêtre redimensionnable
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);                //openGL 3.2
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);

        glfwWindow = glfwCreateWindow(width, height, title, NULL, NULL);
        if (glfwWindow == NULL)
        {
            throw new RuntimeException("impossible de créer la fenêtre GLFW");
        }

        //pour redimensionner la fenêtre et le viewport d'opengl
        glfwSetFramebufferSizeCallback(glfwWindow, (window, width, height) ->
        {
            this.width = width;
            this.height = height;
            this.setResized(true);
        });

        //ferme la fenêtre en appuyant sur échap
        glfwSetKeyCallback(glfwWindow, (window, key, scancode, action, mods) ->
        {
            if (key == GLFW_KEY_ESCAPE && action == GLFW_RELEASE)
            {
                glfwSetWindowShouldClose(window, true);
            }
        });

        //centrage de la fenêtre
        GLFWVidMode vidMode = glfwGetVideoMode(glfwGetPrimaryMonitor());
        glfwSetWindowPos(
                glfwWindow,
                (vidMode.width() - width) / 2,
                (vidMode.height() - height) / 2
        );

        //activer openGL
        glfwMakeContextCurrent(glfwWindow);

        if (isvSync())
        {
            glfwSwapInterval(1);
        }

        glfwShowWindow(glfwWindow);

        GL.createCapabilities();

        glClearColor(0.0f, 0.0f, 0.0f, 0.0f);

        fps = new FPSCounter();
        fps.start();
    }

    public boolean windowShouldClose()
    {
        return glfwWindowShouldClose(glfwWindow);
    }

    public void setClearColor(float r, float g, float b, float alpha)
    {
        glClearColor(r, g, b, alpha);
    }

    public boolean isKeyPressed(int keyCode)
    {
        return glfwGetKey(glfwWindow, keyCode) == GLFW_PRESS;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public String getTitle() {
        return title;
    }

    public boolean isResized()
    {
        return resized;
    }

    public boolean isvSync() {
        return vSync;
    }

    public void setResized(boolean resized)
    {
        this.resized = resized;
    }

    public void setvSync(boolean vSync)
    {
        this.vSync = vSync;
    }

    public void update()
    {
        if (isResized())
        {
            glViewport(0, 0, width, height);
            setResized(false);
        }

        glfwSwapBuffers(glfwWindow);
        glfwPollEvents();

        fps.frame();
        glfwSetWindowTitle(glfwWindow, this.title + " | FPS:" + fps.get());
    }

}

File: openGL/Engine/Window.cpp
#include "Window.h"


Window::Window(int width, int height, std::string title, bool vSync)
{
	this->width = width;
	this->height = height;
	this->title = title;
	this->resized = false;
	this->vSync = vSync;
}

int Window::init()
{
	if (!glfwInit())
	{
		fprintf(stderr, "impossible d'initaliser l'application");
		return -1;
	}
	glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
	glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
	glfwWindowHint(GLFW_SAMPLES, 2);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3.);
	glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

	this->glfwHandle = glfwCreateWindow(this->width, this->height, this->title.c_str(), NULL, NULL);
	if (glfwHandle == NULL)
	{
		fprintf(stderr, "impossible de créer la fenêtre GLFW");
		glfwTerminate();
		return -1;
	}

	glfwSetInputMode(glfwHandle, GLFW_STICKY_KEYS, GL_TRUE);
	glfwSetFramebufferSizeCallback(glfwHandle, [](GLFWwindow* glfwWindow, int width, int height)
		{
			Window* w = Engine::instance->getWindow();
			w->setWidth(width);
			w->setHeight(height);
			w->setResized(true);
		});

	glfwMakeContextCurrent(glfwHandle);

	//glewExperimental = true;
	//if (glewInit() != GLEW_OK)
	//{
		//fprintf(stderr, "impossible d'initialiser GLEW");
		//glfwTerminate();
		//return -1;
	//}

	if (!gladLoadGL())
	{
		throw new std::exception("Impossible d'initialiser GLAD");
	}

	if (isvSync())
	{
		glfwSwapInterval(1);
	}

	glfwShowWindow(glfwHandle);

	glClearColor(0.0f, 0.0f, 0.0f, 0.0f);

	initFPSCounter();
	return 0;
}

bool Window::shouldClose()
{
	return glfwWindowShouldClose(glfwHandle);
}

void Window::update()
{
	if (isResized())
	{
		glViewport(0, 0, this->width, this->height);
		setResized(false);
	}

	glfwSwapBuffers(glfwHandle);
	glfwPollEvents();

	displayFPS(glfwHandle);
}

void Window::setShouldClose(bool value)
{
	glfwSetWindowShouldClose(glfwHandle, GL_TRUE);
}

void Window::setClearColor(float r, float g, float b, float alpha)
{
	glClearColor(r, g, b, alpha);
}


bool Window::isKeyPressed(int keyCode)
{
	return glfwGetKey(glfwHandle, keyCode) == GLFW_PRESS;
}

int Window::getWidth()
{
	return this->width;
}

int Window::getHeight()
{
	return this->height;
}

std::string Window::getTitle()
{
	return this->title;
}

void Window::setResized(bool value)
{
	this->resized = value;
}

void Window::setWidth(int value)
{
	this->width = width;
}

void Window::setHeight(int value)
{
	this->height = height;
}

void Window::setvSync(bool value)
{
	this->vSync = value;
}

bool Window::isResized()
{
	return this->resized;
}


bool Window::isvSync()
{
	return this->vSync;
}